mdn/developer-portal
# 1017: Automatic content ingestion

This changeset introduces support for automatically adding draft content to the CMS and pinging moderators to approve it.

It pulls data from RSS or Atom feeds and comes with a data migration that automatically configures an initial set of Blogs and YouTube feeds to pull from:

<img width="513" alt="Screenshot 2020-01-14 at 14 09 51" src="https://user-images.githubusercontent.com/101457/72351070-90987300-36d7-11ea-932a-5b5fcbd5b4ad.png">

The background scheduler runs two tasks every two hours (but not at the same time - 30 mins apart), which will attempt to ingest videos or articles respectively. If the relevant feeds show that there's a new item **created** since the last time the feed was checked, it is added to the CMS as draft content and the Moderators are emailed to notify them about this:

```
-------------------------------------------------------------------------------
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: The page "Mozilla Announces Deal to Bring Firefox Reality to Pico
 Devices" has
 been submitted for moderation
From: None
To: notsjalim@mozilla.com
Date: Mon, 13 Jan 2020 11:19:35 -0000
Message-ID: <157891437597.201.13313017575110632161@de69f79bfd63>
Auto-Submitted: auto-generated

Hello Editor,



The page "Mozilla Announces Deal to Bring Firefox Reality to Pico Devices" has been submitted for moderation by Portal Bot.

You can preview the page here: http://localhost:8000/admin/pages/moderation/359/preview/
You can edit the page here: http://localhost:8000/admin/pages/151/edit/



Edit your notification preferences here: http://localhost:8000/admin/account/notification_preferences/
```

The Moderator can then preview, accept or reject the item. Note that rejecting it will still retain the draft version, so it's not a terminal action.

<img width="1347" alt="Screenshot 2020-01-13 at 12 34 14" src="https://user-images.githubusercontent.com/101457/72352502-4664c100-36da-11ea-8f87-e6753e8e72d1.png">

Note that because we're pulling in data from various sources, we need to be sure that the slugs of the items are still unique even if two things have the same title, so we append a hash of the data payload extracted from the feed (this payload includes contains the title, authors and a timestamp, so is fairly reliable for resulting in a unique hash). eg `/videos/we-made-another-channel-35017c1cfa13`

(Related issue #1017)

## How to test

* Ensure your local stack is configured for AWS so that it can save Images to S3 (or that you're sure it can save Images to local disk instead of needing S3)
* When booting up, confirm the scheduled ingestion tasks are registered to run two-hourly

```
scheduler_1  | <ScheduleEntry: run-publish-scheduled-command-every-hour developerportal.apps.taskqueue.tasks.publish_scheduled_pages() <crontab: 55 * * * * (m/h/d/dM/MY)>
scheduler_1  | <ScheduleEntry: selectively-purge-cdn-every-night developerportal.apps.taskqueue.tasks.selectively_invalidate_cdn() <crontab: 15 0 * * * (m/h/d/dM/MY)>
scheduler_1  | <ScheduleEntry: ingest-videos-every-two-hours developerportal.apps.ingestion.tasks.ingest_videos() <crontab: 37 */2 * * * (m/h/d/dM/MY)>
scheduler_1  | <ScheduleEntry: ingest-articles-every-two-hours developerportal.apps.ingestion.tasks.ingest_articles() <crontab: 17 */2 * * * (m/h/d/dM/MY)>
scheduler_1  | <ScheduleEntry: celery.backend_cleanup celery.backend_cleanup() <crontab: 0 4 * * * (m/h/d/dM/MY)>
scheduler_1  | [2020-01-14 13:31:55,252: DEBUG/MainProcess] beat: Ticking with max interval->5.00 minutes
```

and that the tasks to ingest articles and videos are also registered at the worker level.

```
...
worker_1     |   . developerportal.apps.ingestion.tasks.ingest_articles
worker_1     |   . developerportal.apps.ingestion.tasks.ingest_videos
...
```

* Set all IngestionConfiguration `last_sync` dates to 2019-12-12
* `docker-compose exec app python manage.py shell`
* `from developerportal.apps.ingestion.tasks import *`
* Execute `ingest_videos()` - you should see "INFO:developerportal.apps.ingestion.tasks:Skipping automatic video ingestion because settings.AUTOMATICALLY_INGEST_CONTENT is False"
* Edit your `local.py` to set `AUTOMATICALLY_INGEST_CONTENT = True`
* Quit the Django shell and re-enter it.
* Execute `ingest_videos()` again - you should see some logging of it happening, as well as emails going to the console saying that new items are ready for moderation (see below for an example of that)
* Flip to locahost:8000/admin/ - you should see a moderation queue
* Flip to localhost:8000/posts/ - confirm nothing in the moderation queue appears yet in the live page
* In the admin, preview a draft/pending Video page (opens a new tab), then close the tab
* In the admin, back on the moderation list, approve that Video
* Confirm it now appears in localhost:8000/posts/ complete with a thumbnail
* Reject a video and confirm it does not appear in the /posts/ page after that (but the Draft page does still exist in Wagtail)

* Now do the same for `ingest_articles()`, which creates ExternalArticles (ie, links to externally hosted content, rather than creating pages inside the CMS that hold the same content) 
* Confirm that previewing them opens a new tab, outside of the CMS;
* approve some, reject others, as above, confirming presence/absence from /posts/ accordingly.
* Re-run the two `ingest_*()` functions and note that duplicates are gracefully handled:
```WARNING:developerportal.apps.ingestion.utils:Problem ingesting article from {'title': 'Mozilla Hacks’ 10 most-read posts of 2019', 'authors': ['Havi Hoffman'], 'url': 'https://hacks.mozilla.org/2019/12/mozilla-hacks-most-read-blog-posts-of-2019/', 'image_url': '', 'timestamp': datetime.datetime(2019, 12, 17, 17, 22, 57, tzinfo=tzlocal()), 'owner': <User: ingestion_user>}: {'slug': ['This slug is already in use']}```

Note that this code is currently running on the dev env at https://developer-portal.dev.mdn.mozit.cloud/, so we'll see how the automatic ingestion works on a schedule
